src/mapgen_carpathian.cpp

```diff
@@ -0,0 +1,451 @@
+/*
+Minetest
+Copyright (C) 2010-2016 paramat, Matt Gregory
+Copyright (C) 2010-2016 kwolekr, Ryan Kwolek <kwolekr@minetest.net>
+Copyright (C) 2017 vlapsley, Vaughan Lapsley <vlapsley@gmail.com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+
+#include "mapgen.h"
+#include "voxel.h"
+#include "noise.h"
+#include "mapblock.h"
+#include "mapnode.h"
+#include "map.h"
+#include "content_sao.h"
+#include "nodedef.h"
+#include "voxelalgorithms.h"
+//#include "profiler.h" // For TimeTaker
+#include "settings.h" // For g_settings
+#include "emerge.h"
+#include "dungeongen.h"
+#include "cavegen.h"
+#include "mg_biome.h"
+#include "mg_ore.h"
+#include "mg_decoration.h"
+#include "mapgen_carpathian.h"
+
+
+FlagDesc flagdesc_mapgen_carpathian[] = {
+	{"caverns", MGCARPATHIAN_CAVERNS},
+	{NULL,      0}
+};
+
+
+///////////////////////////////////////////////////////////////////////////////
+
+
+MapgenCarpathian::MapgenCarpathian(
+		int mapgenid, MapgenCarpathianParams *params, EmergeManager *emerge)
+	: MapgenBasic(mapgenid, params, emerge)
+{
+	this->spflags          = params->spflags;
+	this->cave_width       = params->cave_width;
+	this->cavern_limit     = params->cavern_limit;
+	this->cavern_taper     = params->cavern_taper;
+	this->cavern_threshold = params->cavern_threshold;
```

Comment:
Since you are using large pseudorandom caves 2 more parameters are needed to be able to shift those caves and the lava they contain vertically if terrain is shifted vertically, see mgv7:
```
	this->large_cave_depth    = params->large_cave_depth;
	this->lava_depth          = params->lava_depth;
```